src/mainloop.rkt

```diff
@@ -330,22 +326,27 @@
   (define original-points (setup-context! vars (or specification prog) precondition repr))
   (run-improve! iters prog specification preprocess original-points repr))
 
-(define (run-improve! expression context pcontext rules iterations #:specification [specification #f])
+(define (run-improve! initial specification context pcontext
+                      rules iterations node-limit)
   (timeline-event! 'preprocess)
-  (define preprocessing (find-preprocessing (or specification expression) context rules))
+  (define-values (simplified preprocessing)
+    (find-preprocessing
+     initial specification context
+     rules node-limit (flag-set? 'setup 'simplify)))
   (timeline-push! 'symmetry (map ~a preprocessing))
   (define pcontext* (preprocess-pcontext context pcontext preprocessing))
   (match-define (and alternatives (cons best _))
-    (mutate! expression iterations pcontext*))
+    (mutate! simplified context pcontext* iterations))
   (timeline-event! 'preprocess)
   (define preprocessing*
     (remove-unnecessary-preprocessing (alt-expr best) context pcontext preprocessing))
   (values alternatives preprocessing*))
 
-(define (mutate! prog iters pcontext)
+(define (mutate! simplified context pcontext iterations)
```

Comment:
Looks like the change here is just to avoid relying on `(*context*)`? I guess that's fine.